src/driver-sinowealth.c

```diff
@@ -0,0 +1,400 @@
+	 * but the glorious software only exposes six
+	 */
+	uint8_t dpi[16];
+	/* DPI/CPI is encoded in the way the PMW3360 sensor accepts it
+	 * value = (DPI - 100) / 100
+	 * If XY are identical, dpi[0-6] contain the sensitivities,
+	 * while in XY independent mode each entry takes two chars for X and Y.
+	 */
+	RGB8 dpi_color[8];
+
+	uint8_t rgb_effect;
+	/* see enum rgb_effect */
+
+	char glorious_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	char glorious_direction;
+
+	RGB8 single_color;
+
+	char breathing_mode;
+	/* 0x40 - brightness (constant)
+	 * 0x1/2/3 - speed
+	 */
+	char breathing_colorcount;
+	/* 7, constant */
+	RGB8 breathing_colors[7];
+
+	char tail_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+
+	char rave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+	RGB8 rave_colors[2];
+
+	char wave_mode;
+	/* 0x10/20/30/40 - brightness
+	 * 0x1/2/3 - speed
+	 */
+
+	char breathing1_mode;
+	/* 0x1/2/3 - speed */
+	RGB8 breathing1_color;
+
+	char unk4;
+	char lift_off_distance;
+	/* 0x1 - 2 mm
+	 * 0x2 - 3 mm
+	 */
+} __attribute__((packed));
+
+struct sinowealth_data {
+	/* this is kinda unnecessary at this time, but all the other drivers do it too ;) */
+	struct sinowealth_config_report config;
+};
+
+
+static int
+sinowealth_raw_to_dpi(int raw)
+{
+	return (raw + 1) * 100;
+}
+
+static int
+sinowealth_dpi_to_raw(int dpi)
+{
+	return dpi / 100 - 1;
+}
+
+static struct ratbag_color
+sinowealth_raw_to_color(RGB8 raw)
+{
+	return (struct ratbag_color) {.red = raw.r, .green = raw.g, .blue = raw.b};
+}
+
+static RGB8
+sinowealth_color_to_raw(struct ratbag_color color)
+{
+	return (RGB8) {.r = color.red, .g = color.green, .b = color.blue};
+}
+
+static int
+sinowealth_read_profile(struct ratbag_profile *profile)
+{
+	struct ratbag_device *device = profile->device;
+	struct sinowealth_data *drv_data = device->drv_data;
+	struct sinowealth_config_report *config = &drv_data->config;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+	int num_dpis = (SINOWEALTH_DPI_MAX - SINOWEALTH_DPI_MIN) / SINOWEALTH_DPI_STEP + 1 + 1;
+	unsigned int dpis[num_dpis];
+	unsigned int hz = 1000; /* TODO */
+	int rc;
+	uint8_t *data;
+
+	uint8_t cmd[6] = {SINOWEALTH_REPORT_ID_CMD, SINOWEALTH_CMD_GET_CONFIG};
+	rc = ratbag_hidraw_set_feature_report(device, SINOWEALTH_REPORT_ID_CMD, cmd, sizeof(cmd));
+	if(rc != sizeof(cmd)) {
+		log_error(device->ratbag, "Error while sending read config command: %d\n", rc);
+		return -1;
+	}
+
+	data = zalloc(SINOWEALTH_CONFIG_SIZE);
+	rc = ratbag_hidraw_get_feature_report(device, SINOWEALTH_REPORT_ID_CONFIG,
+					      data, SINOWEALTH_CONFIG_SIZE);
+	/* The GET_FEATURE report length has to be 520, but the actual data returned is less */
+	if (rc < sizeof(*config)) {
+		log_error(device->ratbag, "Could not read device configuration: %d\n", rc);
+		return -1;
+	}
+
+	memcpy(config, data, sizeof(*config));
+	free(data);
```

Comment:
Can you make `sinowealth_config_report` a union and avoid `memcpy` and friends?

Example: 
https://github.com/libratbag/libratbag/blob/02575a90083b6cc3d5dd1a2d510acc34c4c24443/src/hidpp20.h#L44
https://github.com/FFY00/wireshark/blob/385c4da96a6110baac5c6c6f512123837568945e/epan/dissectors/packet-usb-hid.c#L180

Reply:
I changed it a little bit, the struct now just has all the padding required to make it work. No memcpy. The firmware in these devices is a little bit weird; it requires the get/set feature reports to carry 520 bytes (including 1 byte for the report ID), but only 131 are actually used (and sent by the device). I assume this was originally just the entire 512 byte EEPROM (plus a header) of the controller they used, but as features got added they split it up into segments or something like that.